pennylane/gradients/general_shift_rules.py

```diff
@@ -126,4 +166,152 @@ def get_shift_rule(frequencies, shifts=None):
+            eigenvalues.
+        shifts (tuple[int or float]): the tuple of shift values. If unspecified, equidistant
+            shifts are assumed. If supplied, the length of this tuple should match the number of
+            given frequencies.
+        order (int): the order of differentiation to compute the shift rule for
+
+    Returns:
+        tuple: a tuple of coefficients and shifts describing the gradient recipe
+        for the parameter-shift method. For parameter :math:`\phi`, the
+        coefficients :math:`c_i` and the shifts :math:`s_i` combine to give a gradient
+        recipe of the following form:
+
+        .. math:: \frac{\partial}{\partial\phi}f = \sum_{i} c_i f(\phi + s_i).
+
+        where :math:`f(\phi) = \langle 0|U(\phi)^\dagger \hat{O} U(\phi)|0\rangle`
+        for some observable :math:`\hat{O}` and some unitary :math:`U(\phi)`.
+
+    Raises:
+        ValueError: if ``frequencies`` is not a list of unique positive values, or if ``shifts``
+            (if specified) is not a list of unique values the same length as ``frequencies``.
+
+    **Examples**
+
+    An example of obtaining the frequencies from generator eigenvalues, and obtaining the
+    parameter shift rule:
+
+    >>> eigvals = (-0.5, 0, 0, 0.5)
+    >>> frequencies = eigvals_to_frequencies(eigvals)
+    >>> generate_shift_rule(frequencies)
+    [[ 0.85355339, -0.85355339, -0.14644661,  0.14644661],
+     [ 0.78539816, -0.78539816,  2.35619449, -2.35619449]]
+
+    An example with explicitly specified shift values:
+
+    >>> frequencies = (1, 2, 4)
+    >>> shifts = (np.pi / 3, 2 * np.pi / 3, np.pi / 4)
+    >>> generate_shift_rule(frequencies, shifts)
+    [[ 3.        , -3.        , -2.09077028,  2.09077028, 0.2186308, -0.2186308 ],
+     [ 0.78539816, -0.78539816,  1.04719755, -1.04719755, 2.0943951, -2.0943951 ]]
+
+    Higher order shift rules (corresponding to the :math:`n`-th derivative of the parameter)
+    can be requested via the ``order`` argument. For example, to extract the
+    second order shift rule for a gate with generator :math:`X/2`:
+
+    >>> eigvals = (0.5, -0.5)
+    >>> frequencies = eigvals_to_frequencies(eigvals)
+    >>> generate_shift_rule(frequencies, order=2)
+    [[-0.5       ,  0.5       ],
+     [ 0.        , -3.14159265]]
+
+    This corresponds to :math:`\frac{\partial^2 f}{\partial phi^2} = \frac{1}{2} \left[f(\phi) - f(\phi-\pi)]`.
+    """
+    frequencies = tuple(f for f in frequencies if f > 0)
+    recipe = _get_shift_rule(frequencies, shifts=shifts)
+
+    if order > 1:
+        all_shifts = []
+        T = frequencies_to_period(frequencies)
+
+        for partial_recipe in itertools.product(recipe.T, repeat=order):
+            c, s = np.stack(partial_recipe).T
+            new_shift = np.mod(sum(s) + 0.5 * T, T) - 0.5 * T
+            all_shifts.append(np.stack([np.prod(c), new_shift]))
+
+        recipe = qml.math.stack(all_shifts).T

```

Comment:
I am not sure what happens here, maybe a comment would help?

Reply:
done!